### ISSUE TITLE
Date filtering change between v0.48.3 and v0.48.4. looks timezone related

### ISSUE THREAD

**Issue by notrom** [metabase/metabase#38248]
### Describe the bug

I upgraded from v0.48.3 to v0.48.4 and I'm now getting different `where` clauses in Metabase generated SQL Server queries and different results. v0.48.4 appears to be incorrect. There have been no changes to config or questions, but I'm getting different results.

This is the datetime field filter in the UI:
![image](https://github.com/metabase/metabase/assets/4504437/81a82209-edfe-4a94-95b9-2383b503b0f6)

## v0.48.3 gave me this:
```sql
WHERE
      (
        "source"."d" >= DateTimeOffsetFromParts(2023, 8, 28, 0, 0, 0, 0, 0, 0, 7)
       
   AND "source"."d" < DateTimeOffsetFromParts(2023, 8, 30, 0, 0, 0, 0, 0, 0, 7)
      )
```

## v0.48.4 gives me this:
```sql
WHERE
      (
        "source"."d" >= DateTimeOffsetFromParts(2023, 8, 28, 0, 0, 0, 0, 13, 0, 7)
       
   AND "source"."d" < DateTimeOffsetFromParts(2023, 8, 30, 0, 0, 0, 0, 13, 0, 7)
      )
```

Database is MS SQL Server.
Metabase report timezone is "Pacific/Auckland" (although this doesn't matter as field is datetime so no tz stored).
Java timezone set to UTC via command line at startup (to give correct behaviour for MS SQL Server datetimes).

### To Reproduce

1. In 0.48.7, using MS SQL datasource, create a question from this SQL:

<details>
  <summary>Hours query, every hour for 27, 28, and 29 August 2023</summary>

  ```sql
select
          cast('2023-08-27 00:00' as datetime) d
        union all
        select
          cast('2023-08-27 01:00' as datetime) d
        union all
        select
          cast('2023-08-27 02:00' as datetime) d
        union all
        select
          cast('2023-08-27 03:00' as datetime) d
        union all
        select
          cast('2023-08-27 04:00' as datetime) d
        union all
        select
          cast('2023-08-27 05:00' as datetime) d
        union all
        select
          cast('2023-08-27 06:00' as datetime) d
        union all
        select
          cast('2023-08-27 07:00' as datetime) d
        union all
        select
          cast('2023-08-27 08:00' as datetime) d
        union all
        select
          cast('2023-08-27 09:00' as datetime) d
        union all
        select
          cast('2023-08-27 10:00' as datetime) d
        union all
        select
          cast('2023-08-27 11:00' as datetime) d
        union all
        select
          cast('2023-08-27 12:00' as datetime) d
        union all
        select
          cast('2023-08-27 13:00' as datetime) d
        union all
        select
          cast('2023-08-27 14:00' as datetime) d
        union all
        select
          cast('2023-08-27 15:00' as datetime) d
        union all
        select
          cast('2023-08-27 16:00' as datetime) d
        union all
        select
          cast('2023-08-27 17:00' as datetime) d
        union all
        select
          cast('2023-08-27 18:00' as datetime) d
        union all
        select
          cast('2023-08-27 19:00' as datetime) d
        union all
        select
          cast('2023-08-27 20:00' as datetime) d
        union all
        select
          cast('2023-08-27 21:00' as datetime) d
        union all
        select
          cast('2023-08-27 22:00' as datetime) d
        union all
        select
          cast('2023-08-27 23:00' as datetime) d
        union all
        select
          cast('2023-08-28 00:00' as datetime) d
        union all
        select
          cast('2023-08-28 01:00' as datetime) d
        union all
        select
          cast('2023-08-28 02:00' as datetime) d
        union all
        select
          cast('2023-08-28 03:00' as datetime) d
        union all
        select
          cast('2023-08-28 04:00' as datetime) d
        union all
        select
          cast('2023-08-28 05:00' as datetime) d
        union all
        select
          cast('2023-08-28 06:00' as datetime) d
        union all
        select
          cast('2023-08-28 07:00' as datetime) d
        union all
        select
          cast('2023-08-28 08:00' as datetime) d
        union all
        select
          cast('2023-08-28 09:00' as datetime) d
        union all
        select
          cast('2023-08-28 10:00' as datetime) d
        union all
        select
          cast('2023-08-28 11:00' as datetime) d
        union all
        select
          cast('2023-08-28 12:00' as datetime) d
        union all
        select
          cast('2023-08-28 13:00' as datetime) d
        union all
        select
          cast('2023-08-28 14:00' as datetime) d
        union all
        select
          cast('2023-08-28 15:00' as datetime) d
        union all
        select
          cast('2023-08-28 16:00' as datetime) d
        union all
        select
          cast('2023-08-28 17:00' as datetime) d
        union all
        select
          cast('2023-08-28 18:00' as datetime) d
        union all
        select
          cast('2023-08-28 19:00' as datetime) d
        union all
        select
          cast('2023-08-28 20:00' as datetime) d
        union all
        select
          cast('2023-08-28 21:00' as datetime) d
        union all
        select
          cast('2023-08-28 22:00' as datetime) d
        union all
        select
          cast('2023-08-28 23:00' as datetime) d
        union all
        select
          cast('2023-08-29 00:00' as datetime) d
        union all
        select
          cast('2023-08-29 01:00' as datetime) d
        union all
        select
          cast('2023-08-29 02:00' as datetime) d
        union all
        select
          cast('2023-08-29 03:00' as datetime) d
        union all
        select
          cast('2023-08-29 04:00' as datetime) d
        union all
        select
          cast('2023-08-29 05:00' as datetime) d
        union all
        select
          cast('2023-08-29 06:00' as datetime) d
        union all
        select
          cast('2023-08-29 07:00' as datetime) d
        union all
        select
          cast('2023-08-29 08:00' as datetime) d
        union all
        select
          cast('2023-08-29 09:00' as datetime) d
        union all
        select
          cast('2023-08-29 10:00' as datetime) d
        union all
        select
          cast('2023-08-29 11:00' as datetime) d
        union all
        select
          cast('2023-08-29 12:00' as datetime) d
        union all
        select
          cast('2023-08-29 13:00' as datetime) d
        union all
        select
          cast('2023-08-29 14:00' as datetime) d
        union all
        select
          cast('2023-08-29 15:00' as datetime) d
        union all
        select
          cast('2023-08-29 16:00' as datetime) d
        union all
        select
          cast('2023-08-29 17:00' as datetime) d
        union all
        select
          cast('2023-08-29 18:00' as datetime) d
        union all
        select
          cast('2023-08-29 19:00' as datetime) d
        union all
        select
          cast('2023-08-29 20:00' as datetime) d
        union all
        select
          cast('2023-08-29 21:00' as datetime) d
        union all
        select
          cast('2023-08-29 22:00' as datetime) d
        union all
        select
          cast('2023-08-29 23:00' as datetime) d
  ```
</details>

3. Create a new GUI question from that saved question
4. Filter by the "d" datetime for the 2 days 28th and 29th
![image](https://github.com/metabase/metabase/assets/4504437/056ceb0b-4929-4c31-96a6-1cc8159052df)

5. Returned data will incorrectly include records with a d datetime in the 27th in addition to the 28th and 29th i.e. 3 days.
![image](https://github.com/metabase/metabase/assets/4504437/bafee186-2676-465b-a69b-d67ef55c18cb)

6. Revert back to v0.48.3, rerun the filtered question and it correctly returns only records on the 28th and 29th
![image](https://github.com/metabase/metabase/assets/4504437/1454b9d8-2808-4561-a7f6-23d1343566a2)

### Expected behavior

Date filtering shouldn't change unexpectedly between releases. Date filters should only return dates in the selected range.

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-NZ",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "Cp1252",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "17.0.2+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "17.0.2",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "17.0.2+8",
    "os.name": "Windows 10",
    "os.version": "10.0",
    "user.language": "en",
    "user.timezone": "UTC"
  },
  "metabase-info": {
    "databases": [
      "sqlserver",
      "postgres",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "12.1"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.6.0"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-01-29",
      "tag": "v0.48.4",
      "hash": "62145b0"
    },
    "settings": {
      "report-timezone": "Pacific/Auckland"
    }
  }
}
```

### Severity

Blocker for me moving to v0.48.4. Date filters are critical for both correctness and building trust with my users.

### Additional context

I've previously spent a long time wrestling with timezones before I realised I had to explicitly set my Java timezone to UTC for date filters to works as expected against my MS SQL Server datetimes (all the dates we use). I have a check on my "always up" dashboard that checks if this is working as I expect.

Metabase report timezone has no impact on this.

I am in New Zealand, currently +13 hours on UTC so this difference is significant.

**Comment by paoliniluis** [comment]
The only fix we shipped about filters was https://github.com/metabase/metabase/pull/38070, but that doesn't seem to alter anything about timezones...

**Comment by notrom** [comment]
This may be related to my old issue #30454 but logged it separately as it seems like something changed on v0.48.4. See also this thread https://discourse.metabase.com/t/date-filter-between-is-incorrectly-giving-me-data-outside-of-the-date-range/35407 

I tried v0.48.5 and I'm still getting this issue.

**Comment by robswanLL** [comment]
Also getting this on v0.48.5, more detail on the discourse thread.

**Comment by lbrdnk** [comment]
Issue is fixed by PR #38642. Apart from new tests contained in the PR, I've checked correctness manually executing reproduction steps.
